plugins/Fable.React.Plugins.fs

```diff
@@ -0,0 +1,108 @@
+            { ThisArg = None
+              Args = args
+              SignatureArgTypes = []
+              HasSpread = false
+              IsJsConstructor = false }
+        Fable.Call(callee, callInfo, Fable.Any, None)
+
+    type MemberInfo(?info: Fable.MemberInfo,
+                    ?isValue: bool) =
+
+        let infoOr f v =
+            match info with
+            | Some i -> f i
+            | None -> v
+
+        let argOrInfoOr arg f v =
+            match arg, info with
+            | Some arg, _ -> arg
+            | None, Some i -> f i
+            | None, None -> v
+
+        interface Fable.MemberInfo with
+            member _.IsValue = argOrInfoOr isValue (fun i -> i.IsValue) false
+            member _.Attributes = infoOr (fun i -> i.Attributes) Seq.empty
+            member _.HasSpread = infoOr (fun i -> i.HasSpread) false
+            member _.IsPublic = infoOr (fun i -> i.IsPublic) true
+            member _.IsInstance = infoOr (fun i -> i.IsInstance) true
+            member _.IsMutable = infoOr (fun i -> i.IsMutable) false
+            member _.IsGetter = infoOr (fun i -> i.IsGetter) false
+            member _.IsSetter = infoOr (fun i -> i.IsSetter) false
+            member _.IsEnumerator = infoOr (fun i -> i.IsEnumerator) false
+            member _.IsMangled = infoOr (fun i -> i.IsMangled) false
+
+    let objValue (k, v): Fable.MemberDecl =
+        {
+            Name = k
+            FullDisplayName = k
+            Args = []
+            Body = v
+            UsedNames = Set.empty
+            Info = MemberInfo(isValue=true)
+        }
+
+    let objExpr kvs =
+        Fable.ObjectExpr(List.map objValue kvs, Fable.Any, None)
+
+
+type ReactComponentAttribute() =
+    inherit MemberDeclarationPluginAttribute()
+    override _.FableMinimumVersion = "3.0"
+    override _.Transform(logger, decl) =
+        if decl.Info.IsValue || decl.Info.IsGetter || decl.Info.IsSetter then
+            logger.LogWarning("Expecting a function for ReactComponent")
+            decl
+        else
+            let propsArg = AstUtil.makeIdent "$props"
+            let renderIdent = AstUtil.cleanFullDisplayName decl.FullDisplayName |> AstUtil.makeIdent
+            let renderBody =
+                ([], decl.Args) ||> List.fold (fun bindings arg ->
+                    let getterKind = Fable.ByKey(Fable.ExprKey(AstUtil.makeStrConst arg.DisplayName))
+                    let getter = Fable.Get(Fable.IdentExpr propsArg, getterKind, Fable.Any, None)
+                    (arg, getter)::bindings)
+                |> List.rev
+                |> fun bindings -> Fable.Let(bindings, decl.Body)
+
+            let propsObj =
+                decl.Args
+                |> List.map (fun arg -> arg.DisplayName, Fable.IdentExpr arg)
+                |> AstUtil.objExpr
+            let body =
+                let createEl = AstUtil.makeCall (AstUtil.makeImport "createElement" "react") [Fable.IdentExpr renderIdent; propsObj]
```

Comment:
@alfonsogarciacaro Although this works, combining the component definition with component creation will _most likely_ cause problems for fast-refresh plugin that provides live-reload capabilities for pure React apps. I need to test it first to make sure but I according the discussions I had with the people working on fast-refresh, I believe we will need to have two Fable plugins: one that rewrites the input arguments into `$prop` declaration _and_ another Fable plugin that rewrites call-site into a `createElement` call:
```fs
[<ReactComponent>]
let hello (name: string) = Html.div ("Hello, " + name)

// somewhere down the line
hello "world"
```
should compile to 
```js
export const Hello = ($props) => {
  const { name } = $props
  return Feliz_Html.div("Hello, " + name)
}

createElement(Hello, { name: "world" })
```
Do you think something like that is possible? 

> Notice the upper case transform, this is also something required by the fast-refresh plugin

Reply:
I see, thanks for letting me know. It should be possible with the CallPlugin. Let me give it a try.

Reply:
Bonus points if we could elminite `$props` altogether and instead spread the arguments when defining the component:
```js
export const Hello = ({ name }) => {
  return Feliz_Html.div("Hello, " + name)
}
```
I think there will also be some edge cases when someone uses `key` and `children` as input arguments where:
 - `key` can be provided on call-site but is not available from inside the component (show warnings when `key` is used?)
 - `children` provided on call-site should be _spread_ (like what `[<ParamList>]` does) as the third argument of `createElement`

But you can leave those for me to work on :wink: maybe I won't have to implement a Feliz specific plugin if this does the correct transforms and I would only implement Feliz specific transforms to simplify the generated javascript

Reply:
Hmm, that's gonna be tricky because we have nothing in Fable AST for object spread, only for spreading the last argument in call (we haven't added it to our Babel AST implementation yet either) and we cannot use `Emit` with the arguments. We would have to detect the pattern either in the Fable > Babel or printing phases but it may not be easy.

About `key` and `children` we should be able to deal with these differently to other arguments, I guess.

Reply:
> Hmm, that's gonna be tricky because we have nothing in Fable AST for object spread, only for spreading the last argument in call (we haven't added it to our Babel AST implementation yet either) and we cannot use Emit with the arguments. We would have to detect the pattern either in the Fable > Babel or printing phases but it may not be easy.

It's alright, it is just cosmetics but definitely a nice to have

Reply:
@Zaid-Ajaj Please check the [latest commit](https://github.com/fable-compiler/fable-react/pull/207/commits/6b81f1e8656fe222e5ff00e43ea65c3f1cc22e82). You need to update both fable tool and Fable.AST but now you can transform the call site too 👍

Reply:
@alfonsogarciacaro one question, how to determine that a `Fable.Type` is a record instance? Right now I have this:
```fs
let isRecord (fableType: Fable.Type) =
    match fableType with
    | Fable.Type.AnonymousRecordType _ -> true
    | _ -> false
```
I want to _not_ rewrite the props when the only input used is a single record instance or anonymous record but I am not sure how to check for records :/

I know it must be must something that has to do with `EntityRef` but I only get a `name` out of it and I cannot check whether that name is actually referring to a record type

Reply:
Ah, yes, sorry. I removed the direct references for entities in the AST (this is done to make serialization easier later). I was thinking we should pass the [Compiler object](https://github.com/fable-compiler/Fable/blob/f37a97e61bb67d0c921a1d3780cb361ff975d9fb/src/Fable.Transforms/Global/Compiler.fs#L54-L66) (or parts of it, as it currently has a dependency to FCS) through the plugin helper, and this is probably a good reason for that [as you can use the compiler to get the entity from the reference](https://github.com/fable-compiler/Fable/blob/f37a97e61bb67d0c921a1d3780cb361ff975d9fb/src/Fable.Transforms/Replacements.fs#L739-L744). I'll do it for the next Fable.AST/fable update 👍

Reply:
That would be great! thanks a lot :pray: 🙇 

I am making quite the progress with these plugins, very cool stuff :heart:
```